res/res_pjsip_authenticator_digest.c

```diff
@@ -131,58 +139,132 @@ static const struct ast_sip_auth *get_auth(void)
+		if (pj_strcmp2(&auth_hdr->credential.common.realm, realm) == 0) {
+			SCOPE_EXIT_RTN_VALUE(auth_hdr, "%s:%s: realm: %s Found header\n",
+				auth_id, src_name, realm);
+		}
+	}
+	SCOPE_EXIT_RTN_VALUE(NULL, "%s:%s: realm: %s No auth header found\n",
+		auth_id, src_name, realm);
+}
+
 /*!
  * \brief Lookup callback for authentication verification
  *
  * This function is called when we call pjsip_auth_srv_verify(). It
  * expects us to verify that the realm and account name from the
- * Authorization header is correct. We are then supposed to supply
- * a password or MD5 sum of credentials.
+ * Authorization header are correct and that we can support the digest
+ * algorithm specified. We are then supposed to supply a password or
+ * password_digest for the algorithm.
+ *
+ * The auth object must have previously been saved to thread-local storage.
  *
  * \param pool A memory pool we can use for allocations
- * \param realm The realm from the Authorization header
- * \param acc_name the user from the Authorization header
- * \param[out] info The credentials we need to fill in
+ * \param param Contains the realm, username, rdata and auth header
+ * \param cred_info The credentials we need to fill in
  * \retval PJ_SUCCESS Successful authentication
  * \retval other Unsuccessful
  */
-static pj_status_t digest_lookup(pj_pool_t *pool, const pj_str_t *realm,
-		const pj_str_t *acc_name, pjsip_cred_info *info)
+static pj_status_t digest_lookup(pj_pool_t *pool,
+	const pjsip_auth_lookup_cred_param *param,
+	pjsip_cred_info *cred_info)
 {
-	const struct ast_sip_auth *auth;
+	const struct ast_sip_auth *auth = get_auth();
+	const char *realm = S_OR(auth->realm, default_realm);
+	const char *creds;
+	const char *auth_name = (auth ? ast_sorcery_object_get_id(auth) : "none");
+	struct pjsip_authorization_hdr *auth_hdr = get_authorization_hdr(auth_name, realm, param->rdata);
+	const pjsip_auth_algorithm *algorithm =
+		ast_sip_auth_get_algorithm_by_iana_name(&auth_hdr->credential.digest.algorithm);
+	const char *src_name = param->rdata->pkt_info.src_name;
+	SCOPE_ENTER(4, "%s:%s:"
+		" srv realm: " PJSTR_PRINTF_SPEC
+		" auth realm: %s"
+		" hdr realm: " PJSTR_PRINTF_SPEC
+		" auth user: %s"
+		" hdr user: " PJSTR_PRINTF_SPEC
+		" algorithm: " PJSTR_PRINTF_SPEC
+		"\n",
+		auth_name, src_name,
+		PJSTR_PRINTF_VAR(param->realm),
+		realm,
+		PJSTR_PRINTF_VAR(auth_hdr->credential.common.realm),
+		auth->auth_user,
+		PJSTR_PRINTF_VAR(param->acc_name),
+		PJSTR_PRINTF_VAR(algorithm->iana_name));
 
-	auth = get_auth();
 	if (!auth) {
-		return PJSIP_SC_FORBIDDEN;
+		/* This can only happen if the auth object was not saved to thread-local storage */
+		SCOPE_EXIT_RTN_VALUE(PJSIP_SC_FORBIDDEN, "%s:%s: No auth object found\n",
+			auth_name, src_name);
 	}
 
 	if (auth->type == AST_SIP_AUTH_TYPE_ARTIFICIAL) {
-		return PJSIP_SC_FORBIDDEN;
+		/*
+		 * This shouldn't happen because this function can only be invoked
```

Comment:
Wouldn't this be the case if we challenge on artificial auth, and they resent with Authorization?

Reply:
There's no scenario where we would actually verify an artificial auth.  We can challenge with an artificial auth if the endpoint only has auth_username for instance, but when the challenge comes back it has a userid so we look that up to find the endpoint and auth object.  Besides, the artificial auth has no credentials to verify against.  This is checked down in digest_check_auth().